src/matrixrtc/CallMembership.ts

```diff
@@ -14,24 +14,93 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+import { EitherAnd } from "matrix-events-sdk/lib/types";
+
 import { MatrixEvent } from "../matrix";
 import { deepCompare } from "../utils";
 import { Focus } from "./focus";
 
 type CallScope = "m.room" | "m.user";
-
 // Represents an entry in the memberships section of an m.call.member event as it is on the wire
-export interface CallMembershipData {
-    application?: string;
+
+// There are two different data interfaces. One for the Legacy types and one complient with MSC4143
+
+// MSC4143 (MatrixRTC) session membership data
+
+export type SessionMembershipData = {
+    application: string;
+    call_id: string;
+    device_id: string;
+
+    foci_active: Focus;
+    foci_preferred: Focus[];
+    created_ts?: number;
+
+    // Application specific data
+    scope?: CallScope;
+};
+
+export const isSessionMembershipData = (data: any): data is SessionMembershipData =>
```

Comment:
I think it's most useful / accurate for `isSessionMembershipData` and `isLegacyCallMembershipData` to be used for distinguishing whether a `CallMembershipData` object is the legacy type or not.  That would let both functions specify a `CallMembershipData` as its parameter, and check only a single field.  This would effectively make it equivalent to an inlined typecheck (which would allow for type narrowing without a type predicate function, but would clog code / be less maintainable, as you pointed out).

The "full" checking functions could then accept an `any` and do the job of checking if an arbitrary object is one of the two types (which is how they're used in practice).

Reply:
I was trying that before but it turned out to be less readable. But you are right this is what it should be.
I still think it is a little bit less readable (in the CallMembership construcor) but it makes more sense like this.